include/benchmark/benchmark_api.h

```diff
@@ -445,15 +449,71 @@ class Benchmark {
   // Equivalent to ThreadRange(NumCPUs(), NumCPUs())
   Benchmark* ThreadPerCpu();
 
+  virtual void Run(State& state) = 0;
+
   // Used inside the benchmark implementation
   struct Instance;
 
- private:
-   BenchmarkImp* imp_;
-   BENCHMARK_DISALLOW_COPY_AND_ASSIGN(Benchmark);
+protected:
+   Benchmark(Benchmark const&);
+   void SetName(const char* name);
+
+private:
+  friend class BenchmarkFamilies;
+  BenchmarkImp* imp_;
+
+  Benchmark& operator=(Benchmark const&);
+};
+
+// The class used to hold all Benchmarks created from static function.
+// (ie those created using the BENCHMARK(...) macros.
+class FunctionBenchmark : public Benchmark {
+public:
+    FunctionBenchmark(const char* name, Function* func)
+        : Benchmark(name), func_(func)
+    {}
+
+    virtual void Run(State& st);
+private:
+    Function* func_;
 };
 
 }  // end namespace internal
+
+// The base class for all fixture tests. Fixture tests are created by
+// first defining a type that derives from ::benchmark::Fixture and then
+// creating/registering the tests using the following macros:
+//
+// * BENCHMARK_F(ClassName, Method)
+// * BENCHMARK_DEFINE_F(ClassName, Method)
+// * BENCHMARK_REGISTER_F(ClassName, Method)
+//
+// For Example:
+//
+// class MyFixture : public benchmark::Fixture {};
+//
+// BENCHMARK_F(MyFixture, FooTest)(benchmark::State& st) {
+//    while (st.KeepRunning()) {
+//      ...
+//    }
+// }
+class Fixture: public internal::Benchmark {
+public:
+    Fixture() : internal::Benchmark("") {}
+
+    virtual void Run(State& st) {
+      this->SetUp();
```

Comment:
it's a shame we can't rely on the constructor/destructor pair to be setup/teardown. however, i agree that that won't work well here.